### ISSUE TITLE
MaxMind Database returning questionable data that doesn't match online demo

### ISSUE THREAD

**Issue by nickwilliams-eventbrite** [maxmind/GeoIP2-python#22]
We're using the GeoIP2 City database with the GeoIP2-python library. We're making about 1,000 lookups per second against the database, spread out over thirty servers. Each server has its own copy of the database, so that's about 34 requests per second per GeoIP2 City database file. We're using prefork UWSGI processes, not actual multithreading. We open a `geoip2.database.Reader` in the parent process (this might be the problem) and use it in worker UWSGI processes.

Last night for about 30 minutes we got a surge of questionable data from our MaxMind database copies that doesn't make any sense. After that 30 minutes the problem went away and hasn't returned yet. I demonstrate this questionable data below in the following format: IP address / location according to MaxMind online demo / location according to our MaxMind database lookup on our server:

107.92.59.253 / "Houston, Texas, US" / "#, Aland Islands"
68.8.132.240 / "San Diego, California, US" / "<ñ´¦, New Zealand"
66.87.83.113 / "Philadelphia, Pennsylvania, US" / "E, France"
66.115.83.131 / "Glasgow, Kentucky, US" / "x, Pitcairn Islands"
23.31.73.65 / "Boulder, Colorado, US" / "ê, Malaysia"
72.35.146.175 / "Bend, Oregon, US" / "(È#ÜK®80, Cook Islands"
151.252.29.139 / "Sweden, Europe" / "ð, Nauru"

What could be causing this? Is this a bug in the MaxMind database Python library? Or is it how we're using it?

**Comment by oschwald** [comment]
It should be fine to open the reader before forking. Do you know what reader mode you are using (C extension, mmap, file, memory)? Was the database file updated int he period?

**Comment by nickwilliams-eventbrite** [comment]
The database file was last updated last week, so that wasn't related.

We are opening the reader like so:

```python
_maxmind = geoip2.database.Reader(settings.MAXMIND_GEOIP2_CITY_DATABASE)
```

Which means it's trying MODE_MMAP_EXT, MODE_MMAP, and MODE_FILE in that order. We don't have the C extension installed (it's on our roadmap). Should I be using MODE_MEMORY? I don't understand the difference between the modes.

**Comment by oschwald** [comment]
There isn't much of an advantage to memory over mmap if your Python supports mmap and most do excluding some special installs such as on Google App Engine. You could test this by trying to import `mmap`.

Were all 30 servers affected at the same time? That seems really bizarre if that was the case.

**Comment by nickwilliams-eventbrite** [comment]
Most of them were affected at the same time, yes. Looks like all 30 but I can't say that with 100% certainty. Just that I can confirm _at least_ 20 were affected. We did have a peak traffic period during that time. Perhaps we reached a saturation point that revealed a threading problem in the library?

We're running Python 2.6 on Ubuntu 10.04 on AWS servers (we're in the process of upgrading to Python 2.7 on Ubuntu 12.04), so it sounds like we probably have mmap available. I don't have access to those servers directly so I can't personally test that, but I'm sure we have it.

If you could tell me a little bit more about how the library reads the underlying file and seeks to particular records, perhaps I could see where a problem might be occurring. In my imagination, I see an index at the beginning of the file, the reader loading the index into memory at startup, then finding the pointer in the index when I lookup is made, and then seeking to that point in the file to return the entire record. If my imagination is correct about that, then multiple processes trying to seek the same file pointer to different positions could create the weird data I'm seeing.

**Comment by oschwald** [comment]
We haven't had any other reports of such failures, although that of course does not eliminate the possibility of a bug. You mention a threading problem but previously you mentioned forking. Is your application multithreaded in addition to the preforking that it does? The library may not be thread safe, particularly with the file mode, although it might be safe for CPython threading with the GIL as it is.

As far as how it works, the code should be pretty straightforward. The file is mmaped. The metadata is read into memory and used to create pointers to the different data sections. Then each lookup uses those pointers (but does not change them) to find the data for the lookup.

**Comment by nickwilliams-eventbrite** [comment]
Our application does not use multithreading. It uses prefork UWSGI multiprocessing.

**Comment by nickwilliams-eventbrite** [comment]
I wrote a complex test case which I'll share below. It does the following:

1. Generate a random list of 500-1000 IP addresses
2. Before multithreading or multiprocessing, lookup each address and store the discovered "City, State" or "City, Country" string for that address.
3. Start 64 threads and, in each thread, create a shuffled copy of the list of IP addresses and lookup each one. If its location string doesn't match the one previously looked up, print an error.
4. Wait for all threads to end (re-join).
5. Start 64 forked processes and, in each process, create a shuffled copy of the list of IP addresses and lookup each one. If its location string doesn't match the one previously looked up, print an error.
6. Wait for all processes to end (re-join).

Upon doing this, I found the following result:

- I could not duplicate the errors we were seeing last night with `MODE_MMAP`.
- I could not duplicate the errors with `MODE_MEMORY`.
- I _could_ duplicate the errors with `MODE_FILE`.

However, we _can_ import mmap in production, so I'm going to look deeper into this.

```python
import copy
import multiprocessing
import random
import socket
import struct
import sys
import threading

from geoip2.errors import AddressNotFoundError

from unittest import TestCase

import geoip2.database
import maxminddb.const


class TestMaxmind(TestCase):

    def setUp(self):
        self.maxmind = geoip2.database.Reader(
            filename=settings.MAXMIND_GEOIP2_CITY_DATABASE,
            mode=maxminddb.const.MODE_MEMORY
        )

    def test_maxmind_threading(self):
        # first, generate 1,000 random IP addresses
        print
        print "Generating..."
        ip_addresses = []
        for i in range(0, 1000):
            ip_addresses.append(
                socket.inet_ntoa(struct.pack('>I', random.randint(1, 0xffffffff)))
            )
        print

        # Now look them all up
        countries = set()
        ip_address_locations = dict()
        print "Initial lookup..."
        for ip_address in ip_addresses:
            try:
                result = self.maxmind.city(ip_address)
            except AddressNotFoundError:
                result = None
            string = TestMaxmind._result_to_string(result)
            if string:
                countries.add(result.country.iso_code)
                ip_address_locations[ip_address] = string

        print "%s results across %s countries." % (len(ip_address_locations), len(countries))
        print countries
        print

        keys = ip_address_locations.keys()

        def lookup(indicator):
            sys.stdout.write(indicator)
            sys.stdout.flush()
            random_keys = copy.copy(keys)
            random.shuffle(random_keys)
            for k in random_keys:
                try:
                    r = self.maxmind.city(k)
                except AddressNotFoundError:
                    r = None
                    print "Failed to lookup %s!" % (k, )
                s = TestMaxmind._result_to_string(r)
                v = ip_address_locations[k]
                if s != v:
                    print "For %s expected %s but got %s!" % (k, v, s)

        # Now test them all again in multiple threads and make sure they still match
        print "Testing multiple threads..."
        threads = []
        for i in range(0, 64):
            t = threading.Thread(target=lookup, args=("t", ))
            t.start()
            threads.append(t)
        for t in threads:
            t.join()
        print
        print "Done."
        print

        # Now test them all again in multiple processes and make sure they still match
        print "Testing multiple processes..."
        processes = []
        for i in range(0, 64):
            p = multiprocessing.Process(target=lookup, args=("f", ))
            p.start()
            processes.append(p)
        for p in processes:
            p.join()
        print
        print "Done."
        print

    @classmethod
    def _result_to_string(cls, result):
        if result:
            if result.country.iso_code == "US" and result.subdivisions.most_specific.name and result.city.name:
                return u"%s, %s" % (result.city.name, result.subdivisions.most_specific.name)
            if result.country.name and result.city.name:
                return u"%s, %s" % (result.city.name, result.country.name)

        return None
```

**Comment by oschwald** [comment]
Strange. Your test passes for me with all three modes on Python 2.6.9. Is both the multiprocess test and the multithreading test failing or just the latter?

**Comment by oschwald** [comment]
I've got it to start throwing invalid database errors using your test.

**Comment by nickwilliams-eventbrite** [comment]
Well I figured out part of it. Our dev environment was using GeoIP2-Python version 2.1.0, but our production environment mistakenly was still on 2.0.0. `mmap` wasn't supported in 2.0.0. So, that's why my local tests could replicate it with `MODE_FILE` but not `MODE_MMAP` while the problem was happening in production with default settings (which I assumed was `MODE_MMAP` but couldn't be because of the version).

So, that said, it looks like the test above confirms that GeoIP2-Python is _**definitely**_ thread- and multiprocessing-safe for `MODE_MEMORY`, `MODE_MMAP`, and `MODE_MMAP_EXT`, but _**definitely not**_ thread- or multiprocessing-safe for `MODE_FILE`.

**Comment by oschwald** [comment]
The file mode was a relatively recent addition. [This branch](https://github.com/maxmind/MaxMind-DB-Reader-python/tree/greg/no-auto-file-mode) should fix the issue, but the locking will make file mode slow on Pythons without `os.pread`. If someone is set on using the file mode in a multiprocess situation, they are probably better off opening the database after forking.

**Comment by oschwald** [comment]
Also, thanks for the test case. It was definitely useful in tracking down the issue.

**Comment by nickwilliams-eventbrite** [comment]
Glad I could help. Great job getting it fixed so quickly!

**Comment by oschwald** [comment]
Closing as this has been fixed in 1.2.0 of maxminddb.
